### PR TITLE
OVN: Host-network to remote pod/service support

### PR DESCRIPTION
**Pull request by sridhargaddam**
This PR implements Host network to remote pod/service support
in OVN route-agent driver.

Fixes issue: https://github.com/submariner-io/submariner/issues/993
Signed-Off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-Authored-By: Miguel Angel Ajo <majopela@redhat.com>

### CODE REVIEW

File: pkg/routeagent_driver/handlers/ovn/host_networking.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	toRemove := endpointSubnets.Difference(currentRuleRemotes)
+
+	err = ovn.programRulesForRemoteSubnets(toRemove, netlink.RuleDel, os.IsNotExist)
+	if err != nil {
+		return errors.Wrapf(err, "error removing routing rule")
+	}
+
+	nextHop, err := ovn.getNextHopOnK8sMgmtIntf()
+	if err != nil {
+		return errors.Wrapf(err, "error removing routing rule")
```
Comment (tpantelis) [inline pkg/routeagent_driver/handlers/ovn/host_networking.go:47]:
```suggestion
		return errors.Wrapf(err, "getNextHopOnK8sMgmtIntf  returned error")
```

File: pkg/routeagent_driver/handlers/ovn/host_networking.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	currentRouteList, err := netlink.RouteList(link, syscall.AF_INET)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error retrieving routes on the link %s", OVNK8sMgmntIntfName)
+	}
+
+	for i := range currentRouteList {
+		klog.V(log.DEBUG).Infof("Processing route %v", currentRouteList[i])
+
+		if currentRouteList[i].Dst == nil || currentRouteList[i].Gw == nil {
+			klog.V(log.DEBUG).Infof("Found nil gw or dst")
```
Comment (mangelajo) [inline pkg/routeagent_driver/handlers/ovn/host_networking.go:113]:
```suggestion
			klog.V(log.DEBUG).Infof("Found nil gw or dst")
```

Let's remove this bit once we verify it's working fine.

File: pkg/routeagent_driver/handlers/ovn/host_networking.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+		for _, subnet := range ovn.localEndpoint.Spec.Subnets {
+			_, localCIDR, _ := net.ParseCIDR(subnet)
+			if currentRouteList[i].Dst == localCIDR && currentRouteList[i].Gw != nil {
+				return &currentRouteList[i].Gw, nil
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("could not find the necessary route")
```
Comment (mangelajo) [inline pkg/routeagent_driver/handlers/ovn/host_networking.go:125]:
```suggestion
	return nil, fmt.Errorf("could not find the route to %q via %q", localCIDR, OVNK8sMgmntIntfName)
```

File: pkg/routeagent_driver/handlers/ovn/south_rules.go [hunk 4]
```diff
@@ -31,21 +31,29 @@ func (ovn *Handler) handleSubnets(subnets []string, ruleFunc func(rule *netlink.
...
+
+		rule.Src = srcCIDR
 	}
 
-	rule := netlink.NewRule()
-	rule.Dst = dstCIDR
-	rule.Src = srcCIDR
-	rule.Table = constants.RouteAgentHostNetworkTableID
+	rule.Table = tableId
+	rule.Priority = tableId
```
Comment (mangelajo) [inline pkg/routeagent_driver/handlers/ovn/south_rules.go:56]:
```suggestion
	rule.Priority = tableId
```

May be we should just use a fixed priority for now, or accept a parameter for it?, setting the prioity as the table ID is confusing :)

**Review by mangelajo — CHANGES_REQUESTED** [review]
@sridhargaddam it looks great, good job. A few nit comments.

**Review by sridhargaddam — COMMENTED** [review]
thanks for reviewing @tpantelis @mangelajo

**Comment by sridhargaddam** [comment]
Verified the PR on an OVN build and all the HostNetworking tests pass

```
•SSSSSSSSSSSSSS
Ran 10 of 30 Specs in 140.283 seconds
SUCCESS! -- 10 Passed | 0 Failed | 0 Pending | 20 Skipped
```

**Review by tpantelis — APPROVED** [review]

File: pkg/routeagent_driver/handlers/ovn/south_rules.go (hunk 4 above)
Comment (mangelajo) [reply pkg/routeagent_driver/handlers/ovn/south_rules.go:56]: Ok, I understand now, but it's confusing. Can we document via comments at least?

I would be more explicit and separate the input of priority & table, even if they will be the same value
in the end.
```
 programRule(dest, src string, tableId, priority int) 
```

Can we do something like this later as a follow up?

**Review by mangelajo — APPROVED** [review]
See comments here: 

https://github.com/submariner-io/submariner/pull/1020/files#r542223756

If we could do a follow up PR, it'd make the code/intent clearer.

File: pkg/routeagent_driver/handlers/ovn/south_rules.go (hunk 4 above)
Comment (sridhargaddam) [reply pkg/routeagent_driver/handlers/ovn/south_rules.go:56]: Sure, we can do that.

**Pull request merged**